New feature - user can create a template by assigning "Template" tag.

CHBrianLiu opened the pull request:
Hi @azu,

Thank you for developing such a good plugin. I am willing to contribute to it!

In previous design, the plugin searches templates under `_Templates` notebook. I enhanced the search so that notes with `Template` tag are considered templates and can be used to create new notes.

If user chooses to create a new note with tagged template, the plugin will remove `Template` tag from new created note because new created note is not a template.

Plugin loading flow chart:
![image](https://user-images.githubusercontent.com/38838945/110023223-e14dae00-7d67-11eb-9146-344ce50d1d22.png)

CHBrianLiu committed:
Change templateDocs variable name to templateBook.

Since db.books.findWithName return a Book object, original templateDocs name
is misleading. Change the name to templateBook to reflect the object type.

CHBrianLiu committed:
Plugin now can find templates by "Template" tag.

While the plugin is loaded, it retrieves all notes within _Templates
notes if any. Then the plugin retrieved all notes with tag name
Template and combine two set of notes together.

CHBrianLiu committed:
Rename docs to templateNotes to reflect the usage.

The templateNotes variable is used to store all notes regarded as
templates.

CHBrianLiu committed:
Remove Template tag from new notes to avoid tag flood.

If "Template" tag is not removed, the new added note would be considered
as a template. However, this new template has no required yaml info
which would cause plugin failed to load.

CHBrianLiu committed:
Remove redundant console output line.

The line is for debug purpose, so remove this line to make logic clearer.

CHBrianLiu committed:
Update doc to explain the usage of Template tag.

azu commented on lib/inkdrop-note-templates.js:
```diff
@@ -81,6 +92,13 @@ module.exports = {
                 updatedAt: +new Date(),
             }
+            // remove Template tag from new note if any.
+            for (let tagIndex = note.tags.length - 1; tagIndex >= 0; tagIndex--) {
```
Probably, we can write this filtering logic like the following.

```js
const templateTag = await db.tags.findWithName(TEMPLATE_TAG_NAME);
const tagsWithoutTemplate = templateTag ? template.tags.filter(tag => tag !== templateTag._id) : [];
const note = {
    ...template.doc,
    _id: db.notes.createId(),
    bookId: currentBookId,
    _rev: undefined,
    title: template.title,
    body: template.body,
    tags: tagsWithoutTemplate,
    createdAt: +new Date(),
    updatedAt: +new Date(),
}
```

https://docs.inkdrop.app/reference/db-tag#findwithnamename

I want to avoid [N+1 problem](https://stackoverflow.com/questions/97197/what-is-the-n1-selects-problem-in-orm-object-relational-mapping).

CHBrianLiu replied:
I totally agree with you on this aspect. Let me modify the implementation.

CHBrianLiu replied:
Consider the following case.
 
1. User has no `Template` tag.
2. The template has two tags: `["2021", "Life"]`.

By executing your suggested code, `templateTag` is `undefined` and `tagsWithoutTemplate` is `[]`. However, I think users may want to preserve the tags instead of clearing all tags. Thus, I make last part of ternary operator as `template.tags` itself without filtering like below code.

```javascript
const tagsWithoutTemplate = templateTag ? template.tags.filter(tag => tag !== templateTag._id) : template.tags;
```

How do you think about this?

CHBrianLiu replied:
There's a relating bug fixing in commit a7c7839. Please check it.

azu replied:
>  I make last part of ternary operator as `template.tags` itself without filtering like below code.

Nice!

azu commented on lib/inkdrop-note-templates.js:
```diff
@@ -4,17 +4,28 @@ const { Liquid } = require('liquidjs');
-        return [];
+    const TEMPLATE_TAG_NAME = "Template";
+
+    let templateNotes = Array();
```
```suggestion
    let templateNotes = [];
```

We can use array literal instead of it.

CHBrianLiu replied:
Okay! Let's use array literal instead!

azu commented on lib/inkdrop-note-templates.js:
```diff
@@ -4,17 +4,28 @@ const { Liquid } = require('liquidjs');
-    if (!docs) {
-        console.log(`Not found docs in ${TEMPLATE_NOTEBOOK_NAME}`);
+
+    if (!templateNotes) {
```
`templateNotes` is `[]` by default. (it will be `true`.)
Check the length of `templateNotes`.
```suggestion
    if (templateNotes.length === 0) {
```

CHBrianLiu replied:
Oops! Such a shame mistake from me. In Python, an empty list is regarded as False in comparison statement. Let me correct it.

azu reviewed (changes requested):
Thanks for PR!

I've commented.

CHBrianLiu committed:
Use array literal instead of Array().

According to azu's suggest, we use array literal instead of class
instantiation.

https://github.com/azu/inkdrop-note-templates/pull/2#discussion_r587919697

CHBrianLiu committed:
Fix empty array checking method.

I used if (!templateNotes) to check if templateNotes array is empty
and escape the function, but in JavaScript, empty array is regarded
as True in such statement.

https://github.com/azu/inkdrop-note-templates/pull/2#discussion_r587921557

CHBrianLiu committed:
Avoid N+1 problem by querying Template tag object before tag filtering.

Original implementation of tag filtering would cause N+1 problem
because plugin queries database for each tag name. Now we query
"Template" tag id before the filtering and the filtering is based on tag
id instead of name.

CHBrianLiu committed:
Bug fixing: Access doc attr then access tags attr due to tags are not set to first level attr.

promises function returns an object with five attributes (line 55)
while tags is an attribute of doc. To access tags attribute, must get
through doc attribute first.

azu reviewed (approved):
LGTM!

azu commented:
@CHBrianLiu Thanks!
I've published [v1.1.0](https://github.com/azu/inkdrop-note-templates/releases/tag/v1.1.0)
